### PR TITLE
unix, windows: add uv_loop_init and uv_loop_close

### PR DESCRIPTION
**Pull request by saghul**
These functions supersede uv_loop_new and uv_loop_delete.
uv_loop_init initialized a user allocated loop and uv_loop_close
removes all associated resources a loop uses after it has finished
execution.

uv_loop_new and uv_loop_delete are now deprecated.

@indutny, @piscisaureus: please review :-)

### CODE REVIEW

File: include/uv.h [hunk 1]
```diff
@@ -261,6 +257,32 @@
...
+ * Initializes a uv_loop_t structure.
+ */
+UV_EXTERN int uv_loop_init(uv_loop_t* loop);
+
+/*
+ * Closes all internal loop resources.  This function must only be called once
+ * the loop has finished it's execution or it will return UV_EBUSY.  After this
+ * function returns the user shall free the memory allocated for the loop.
+ */
+UV_EXTERN uv_loop_close(uv_loop_t* loop);
```
Comment (retep998) [inline include/uv.h]: Shouldn't this have a return type?

Comment (saghul) [inline include/uv.h]: Ups, why didn't you tell me Mr Compiler?!

Comment (indutny) [inline include/uv.h]: Because this is C :) That's how people was writing stuff before return types.

**Comment by saghul** [comment]
Note to self: look into https://github.com/joyent/libuv/issues/435, I might need to solve that one otherwise uv_loop_close will always fail on Windows.

File: src/unix/loop.c [hunk 2]
```diff
@@ -45,6 +45,31 @@ uv_loop_t* uv_default_loop(void) {
...
+  if (!QUEUE_EMPTY(&(loop)->active_reqs))
+    return -EBUSY;
+  QUEUE_FOREACH(q, &loop->handle_queue) {
+    h = QUEUE_DATA(q, uv_handle_t, handle_queue);
+    if (!(h->flags & UV__HANDLE_INTERNAL))
+      return -EBUSY;
+  }
+  uv__loop_close(loop);
+#ifndef NDEBUG
+  memset(loop, -1, sizeof(*loop));
```
Comment (indutny) [inline src/unix/loop.c:66]: Huh? `-1` ?

Comment (indutny) [inline src/unix/loop.c:66]: Ah, that's how it was.

**Comment by saghul** [comment]
Force pushed, fixing some nasty problem in Windows: initializing libuv is kind of important :-) Tests pass on all platforms now (on Windows there are some known failures, this PR doesn't introduce new failures).

Reviewers, please review!

**Comment by indutny** [comment]
If it builds and runs on win and unixes, and if tests are passing - LGTM.

**Comment by saghul** [comment]
Thanks @indutny, landed it!

**Comment by yann2192** [comment]
@saghul uv_loop_close always returns -EBUSY when I use signals : http://pastebin.com/ZqA2hhZG

**Comment by indutny** [comment]
@yann2192 perhaps you should call `uv_close` on `uv_signal_t` first?

**Comment by saghul** [comment]
@yann2192 what @indutny said. Stopping is not enough. For the loop to properly close all handles need to be uv_close'd first.

**Comment by saghul** [comment]
This works: https://gist.github.com/9504483

**Comment by yann2192** [comment]
Oh, I didn't know that uv_signal_t requires uv_close, my bad.

**Comment by saghul** [comment]
On 03/12/2014 11:41 AM, Yann GUIBET wrote:

> Oh, I didn't know that uv_signal_t requires uv_close, my bad.

No worries. All handles do, actually.

## 

Saúl Ibarra Corretgé
bettercallsaghul.com

**Pull request merged**